src/test/scala/cognite/spark/v1/FileContentRelationTest.scala

```diff
@@ -0,0 +1,160 @@
+      mimeType = mimeType,
+    )
+    for {
+      existingFile <- writeClient.files.retrieveByExternalId(externalId).attempt
+
+      // delete file if it was created but the actual file wasn't uploaded so we can get an uploadUrl
+      _ <- existingFile match {
+        case Right(value) if !value.uploaded => writeClient.files.deleteByExternalIds(Seq(externalId))
+        case _ => IO.pure(())
+      }
+
+      file <- existingFile match {
+        case Right(value) if value.uploaded => IO.pure(value)
+        case _ =>
+          writeClient.files.create(Seq(toCreate)).map(_.headOption.getOrElse(throw new IllegalStateException("could not upload file")))
+      }
+
+      _ <- {
+        if (!file.uploaded) {
+          val backend = Resource.make(IO(HttpURLConnectionBackend()))(b => IO(b.close()))
+          val request = basicRequest.put(uri"${file.uploadUrl.getOrElse(throw new MalformedURLException("bad url"))}")
+            .body(generateNdjsonData)
+          backend.use { backend =>
+            IO {
+              val response = request.send(backend)
+              if (response.code.isSuccess) {
+                println(s"NDJSON content uploaded successfully to ${file.uploadUrl}")
+              } else {
+                throw new Exception(s"Failed to upload content: ${response.statusText}")
+              }
+            }
+          }
+        } else {
+          IO.pure(())
+        }
+      }
+    } yield ()
+  }
+
+  "file content transformations" should "read from a ndjson file" in {
+    val sourceDf: DataFrame = dataFrameReaderUsingOidc
+      .useOIDCWrite
+      .option("type", "filecontent")
+      .option("externalId", fileExternalId)
+      .load()
+    sourceDf.createOrReplaceTempView("fileContent")
+    val result = spark.sqlContext.sql(s"select * from filecontent").collect()
+    result.map(_.toSeq.toList) should contain theSameElementsAs
+      Array(
+        List(30, "Alice", null),
+        List(25, "Bob", null),
+        List(35, "Charlie", null),
+        List(35, "Charlie2", "test")
+      )
+  }
+
+  it should "fail with a sensible error if the mimetype is wrong" in {
+    val exception = sparkIntercept {
+      val sourceDf: DataFrame = dataFrameReaderUsingOidc
+        .useOIDCWrite
+        .option("type", "filecontent")
+        .option("externalId", fileWithWrongMimeTypeExternalId)
+        .load()
+      sourceDf.createOrReplaceTempView("fileContent")
+      spark.sqlContext.sql(s"select * from filecontent").collect()
+    }
+    assert(exception.getMessage.contains("Wrong mimetype. Expects application/jsonlines"))
+  }
+
+  it should "infer the schema" in {
+    val sourceDf: DataFrame = dataFrameReaderUsingOidc
+      .useOIDCWrite
+      .option("type", "filecontent")
+      .option("inferSchema", value = true)
+      .option("externalId", fileExternalId)
+      .load()
+    sourceDf.createOrReplaceTempView("fileContent")
+    sourceDf.schema.fields should contain allElementsOf Seq(
```

Comment:
so looking at this test it will try to deduce schemas inside the lines. If we want to keep that, we still should have at least opt-out of schema inference (like in raw)

Reply:
Is it opt out in raw? I though that was a hack ^^

Reply:
opt-out definitely is in place and documented https://docs.cognite.com/cdf/integration/guides/transformation/write_sql_queries/#avoid-schema-inference the choice&errors are not always obvious to users and we don't have a solid recommendation for one or another, so it's not perfect. It also requires SQL rewrite (it may potentially be automated, the column name<>get_json_object part, but still the casting part needs to be encoded somehow so still needs more info to be present in SQL compared to inference-on mode)

Reply:
For some cases no-inference is required to make transformation work despite being less user-friendly (well, more user-friendly in one aspect when there are type conflicts that can be handled via json rather than throwing)

Reply:
I think I found how to make it work :)

Reply:
So, it should now be optional to infer the schema. Should I in addition attempt to flatten the schema? It should be possible as well